File: cargo_bot/bot/handlers/admin/registry.py
"""Реестр сущностей админ-панели.

Каждая сущность описывается декларативно: модель + список полей.
Универсальный CRUD-движок (crud.py) по этому описанию строит списки,
карточки, добавление, редактирование, удаление и вкл/выкл — поэтому
новые управляемые справочники добавляются парой строк.
"""
from dataclasses import dataclass, field as dc_field
from typing import Any, Callable

from bot.models import (
    Admin, Category, Coefficient, Contact, Currency, DeliveryType, ExchangeRate,
    ExtraService, Faq, Setting, Tariff, Warehouse,
)


@dataclass
class Field:
    name: str
    label: str
    ftype: str = "str"  # str | text | int | dec | bool | fk | choice | currency
    required: bool = True
    fk: str | None = None  # ключ сущности для fk-выбора
    choices: list[tuple[str, str]] = dc_field(default_factory=list)  # (значение, подпись)
    hint: str = ""


@dataclass
class Entity:
    key: str
    title: str
    title_one: str
    model: type
    fields: list[Field]
    order_by: list[Any] = dc_field(default_factory=list)
    row: Callable[[Any], str] = str
    can_add: bool = True
    can_delete: bool = True
    per_page: int = 8

    @property
    def has_active(self) -> bool:
        return "is_active" in self.model.__table__.columns


ENTITIES: dict[str, Entity] = {}


def _reg(entity: Entity) -> None:
    ENTITIES[entity.key] = entity


_reg(Entity(
    key="trf", title="💰 Тарифы", title_one="Тариф", model=Tariff,
    order_by=[Tariff.delivery_type_id, Tariff.id],
    fields=[
        Field("name", "Название"),
        Field("delivery_type_id", "Тип доставки", ftype="fk", fk="dtp"),
        Field("category_id", "Категория товара", ftype="fk", fk="cat", required=False,
              hint="Пропустить = общий тариф для всех категорий"),
        Field("price_per_kg", "Стоимость за 1 кг", ftype="dec"),
        Field("min_order_cost", "Минимальная стоимость заказа", ftype="dec"),
        Field("min_weight", "Минимальный оплачиваемый вес, кг", ftype="dec"),
        Field("currency_code", "Валюта расчёта", ftype="currency"),
        Field("delivery_time", "Срок доставки", required=False, hint="Например: 7–12 дней"),
    ],
))

_reg(Entity(
    key="dtp", title="🚚 Типы доставки", title_one="Тип доставки", model=DeliveryType,
    order_by=[DeliveryType.sort_order],
    fields=[
        Field("name", "Название"),
        Field("emoji", "Эмодзи", required=False),
        Field("sort_order", "Порядок сортировки", ftype="int"),
    ],
))

_reg(Entity(
    key="cat", title="🗂 Категории товаров", title_one="Категория", model=Category,
    order_by=[Category.sort_order],
    fields=[
        Field("name", "Название"),
        Field("sort_order", "Порядок сортировки", ftype="int"),
    ],
))

_reg(Entity(
    key="cur", title="💱 Валюты", title_one="Валюта", model=Currency,
    order_by=[Currency.id],
    fields=[
        Field("code", "Код (например USD)"),
        Field("name", "Название"),
        Field("symbol", "Символ", required=False),
        Field("is_base", "Базовая валюта?", ftype="bool"),
    ],
))

_reg(Entity(
    key="rate", title="📈 Курсы валют", title_one="Курс валюты", model=ExchangeRate,
    order_by=[ExchangeRate.currency_id],
    fields=[
        Field("currency_id", "Валюта", ftype="fk", fk="cur"),
        Field("rate", "Курс к TJS (1 валюта = X TJS)", ftype="dec"),
    ],
))

_reg(Entity(
    key="cof", title="🧮 Коэффициенты", title_one="Коэффициент", model=Coefficient,
    order_by=[Coefficient.code, Coefficient.id],
    fields=[
        Field("name", "Название"),
        Field("code", "Тип", ftype="choice", choices=[
            ("volumetric", "Объёмный вес (кг/м³)"),
            ("density", "Плотный груз (множитель цены)"),
            ("custom", "Другой"),
        ]),
        Field("delivery_type_id", "Тип доставки", ftype="fk", fk="dtp", required=False,
              hint="Пропустить = для всех типов"),
        Field("category_id", "Категория", ftype="fk", fk="cat", required=False,
              hint="Пропустить = для всех категорий"),
        Field("value", "Значение", ftype="dec"),
    ],
))

_reg(Entity(
    key="svc", title="➕ Доп. услуги / скидки / наценки", title_one="Услуга/скидка",
    model=ExtraService, order_by=[ExtraService.kind, ExtraService.id],
    fields=[
        Field("name", "Название"),
        Field("kind", "Вид", ftype="choice", choices=[
            ("fee", "Сбор (упаковка, страховка...)"),
            ("discount", "Скидка"),
            ("markup", "Наценка"),
        ]),
        Field("calc_type", "Способ расчёта", ftype="choice", choices=[
            ("fixed", "Фиксированная сумма"),
            ("percent", "% от стоимости"),
            ("per_kg", "Сумма за каждый кг"),
        ]),
        Field("value", "Значение", ftype="dec"),
        Field("delivery_type_id", "Тип доставки", ftype="fk", fk="dtp", required=False,
              hint="Пропустить = для всех типов"),
        Field("category_id", "Категория", ftype="fk", fk="cat", required=False,
              hint="Пропустить = для всех категорий"),
    ],
))

_reg(Entity(
    key="faq", title="❓ FAQ", title_one="Вопрос", model=Faq,
    order_by=[Faq.sort_order],
    fields=[
        Field("question", "Вопрос"),
        Field("answer", "Ответ", ftype="text"),
        Field("sort_order", "Порядок сортировки", ftype="int"),
    ],
))

_reg(Entity(
    key="whs", title="📍 Склады", title_one="Склад", model=Warehouse,
    order_by=[Warehouse.sort_order],
    fields=[
        Field("name", "Название"),
        Field("address", "Адрес", ftype="text"),
        Field("note", "Примечание", ftype="text", required=False),
        Field("sort_order", "Порядок сортировки", ftype="int"),
    ],
))

_reg(Entity(
    key="cnt", title="☎ Контакты", title_one="Контакт", model=Contact,
    order_by=[Contact.sort_order],
    fields=[
        Field("title", "Название (Телефон, Telegram...)"),
        Field("value", "Значение", ftype="text"),
        Field("sort_order", "Порядок сортировки", ftype="int"),
    ],
))

_reg(Entity(
    key="adm", title="👮 Администраторы", title_one="Администратор", model=Admin,
    order_by=[Admin.id],
    fields=[
        Field("tg_id", "Telegram ID", ftype="int"),
        Field("name", "Имя", required=False),
    ],
))

_reg(Entity(
    key="set", title="⚙️ Настройки", title_one="Настройка", model=Setting,
    order_by=[Setting.key],
    fields=[
        Field("key", "Ключ"),
        Field("value", "Значение", ftype="text"),
        Field("description", "Описание", ftype="text", required=False),
    ],
))

File: cargo_bot_php/src/Handler/Admin/Registry.php
<?php

namespace CargoBot\Handler\Admin;

use CargoBot\Util\Format;

/**
 * Реестр управляемых справочников админ-панели.
 *
 * Каждая сущность описана декларативно: таблица + список полей.
 * Универсальный движок (CrudHandler) по этому описанию строит списки,
 * карточки, добавление, редактирование, вкл/выкл и удаление.
 * Чтобы добавить новый справочник, достаточно описать его здесь.
 *
 * Типы полей: str | text | int | dec | bool | fk | choice | currency
 */
class Registry
{
    /** @return array<string,array<string,mixed>> */
    public static function all(): array
    {
        $e = [];

        $e['trf'] = [
            'title' => '💰 Тарифы',
            'one'   => 'Тариф',
            'table' => 'tariffs',
            'order' => 'delivery_type_id, id',
            'has_active' => true,
            'fields' => [
                ['name' => 'name', 'label' => 'Название', 'type' => 'str'],
                ['name' => 'delivery_type_id', 'label' => 'Тип доставки', 'type' => 'fk', 'fk' => 'dtp'],
                ['name' => 'category_id', 'label' => 'Категория товара', 'type' => 'fk', 'fk' => 'cat',
                 'required' => false, 'hint' => 'Пропустить = общий тариф для всех категорий'],
                ['name' => 'price_per_kg', 'label' => 'Стоимость за 1 кг', 'type' => 'dec'],
                ['name' => 'min_order_cost', 'label' => 'Минимальная стоимость заказа', 'type' => 'dec'],
                ['name' => 'min_weight', 'label' => 'Минимальный оплачиваемый вес, кг', 'type' => 'dec'],
                ['name' => 'currency_code', 'label' => 'Валюта расчёта', 'type' => 'currency'],
                ['name' => 'delivery_time', 'label' => 'Срок доставки', 'type' => 'str',
                 'required' => false, 'hint' => 'Например: 7–12 дней'],
            ],
            'row' => fn(array $r) => $r['name'] . ' — ' . Format::money($r['price_per_kg']) . ' ' . $r['currency_code'] . '/кг',
        ];

        $e['dtp'] = [
            'title' => '🚚 Типы доставки',
            'one'   => 'Тип доставки',
            'table' => 'delivery_types',
            'order' => 'sort_order, id',
            'has_active' => true,
            'fields' => [
                ['name' => 'name', 'label' => 'Название', 'type' => 'str'],
                ['name' => 'emoji', 'label' => 'Эмодзи', 'type' => 'str', 'required' => false],
                ['name' => 'sort_order', 'label' => 'Порядок сортировки', 'type' => 'int'],
            ],
            'row' => fn(array $r) => trim(($r['emoji'] ?? '') . ' ' . $r['name']),
        ];

        $e['cat'] = [
            'title' => '🗂 Категории товаров',
            'one'   => 'Категория',
            'table' => 'categories',
            'order' => 'sort_order, id',
            'has_active' => true,
            'fields' => [
                ['name' => 'name', 'label' => 'Название', 'type' => 'str'],
                ['name' => 'sort_order', 'label' => 'Порядок сортировки', 'type' => 'int'],
            ],
            'row' => fn(array $r) => (string) $r['name'],
        ];

        $e['cur'] = [
            'title' => '💱 Валюты',
            'one'   => 'Валюта',
            'table' => 'currencies',
            'order' => 'id',
            'has_active' => true,
            'fields' => [
                ['name' => 'code', 'label' => 'Код (например USD)', 'type' => 'str'],
                ['name' => 'name', 'label' => 'Название', 'type' => 'str'],
                ['name' => 'symbol', 'label' => 'Символ', 'type' => 'str', 'required' => false],
                ['name' => 'is_base', 'label' => 'Базовая валюта?', 'type' => 'bool'],
            ],
            'row' => fn(array $r) => $r['code'] . ' — ' . $r['name'],
        ];

        $e['rate'] = [
            'title' => '📈 Курсы валют',
            'one'   => 'Курс валюты',
            'table' => 'exchange_rates',
            'order' => 'currency_id, id',
            'has_active' => false,
            'stamp_updater' => true,
            'fields' => [
                ['name' => 'currency_id', 'label' => 'Валюта', 'type' => 'fk', 'fk' => 'cur'],
                ['name' => 'rate', 'label' => 'Курс к TJS (1 валюта = X TJS)', 'type' => 'dec'],
            ],
            'row' => fn(array $r) => 'Курс #' . $r['id'] . ' = ' . Format::num($r['rate']),
        ];

        $e['cof'] = [
            'title' => '🧮 Коэффициенты',
            'one'   => 'Коэффициент',
            'table' => 'coefficients',
            'order' => 'code, id',
            'has_active' => true,
            'fields' => [
                ['name' => 'name', 'label' => 'Название', 'type' => 'str'],
                ['name' => 'code', 'label' => 'Тип', 'type' => 'choice', 'choices' => [
                    'volumetric' => 'Объёмный вес (кг/м³)',
                    'density'    => 'Плотный груз (множитель цены)',
                    'custom'     => 'Другой',
                ]],
                ['name' => 'delivery_type_id', 'label' => 'Тип доставки', 'type' => 'fk', 'fk' => 'dtp',
                 'required' => false, 'hint' => 'Пропустить = для всех типов'],
                ['name' => 'category_id', 'label' => 'Категория', 'type' => 'fk', 'fk' => 'cat',
                 'required' => false, 'hint' => 'Пропустить = для всех категорий'],
                ['name' => 'value', 'label' => 'Значение', 'type' => 'dec'],
            ],
            'row' => fn(array $r) => $r['name'] . ' [' . $r['code'] . '] = ' . Format::num($r['value']),
        ];

        $e['svc'] = [
            'title' => '➕ Доп. услуги / скидки / наценки',
            'one'   => 'Услуга или скидка',
            'table' => 'extra_services',
            'order' => 'kind, id',
            'has_active' => true,
            'fields' => [
                ['name' => 'name', 'label' => 'Название', 'type' => 'str'],
                ['name' => 'kind', 'label' => 'Вид', 'type' => 'choice', 'choices' => [
                    'fee'      => 'Сбор (упаковка, страховка...)',
                    'discount' => 'Скидка',
                    'markup'   => 'Наценка',
                ]],
                ['name' => 'calc_type', 'label' => 'Способ расчёта', 'type' => 'choice', 'choices' => [
                    'fixed'   => 'Фиксированная сумма',
                    'percent' => '% от стоимости',
                    'per_kg'  => 'Сумма за каждый кг',
                ]],
                ['name' => 'value', 'label' => 'Значение', 'type' => 'dec'],
                ['name' => 'delivery_type_id', 'label' => 'Тип доставки', 'type' => 'fk', 'fk' => 'dtp',
                 'required' => false, 'hint' => 'Пропустить = для всех типов'],
                ['name' => 'category_id', 'label' => 'Категория', 'type' => 'fk', 'fk' => 'cat',
                 'required' => false, 'hint' => 'Пропустить = для всех категорий'],
            ],
            'row' => function (array $r) {
                $sign = $r['kind'] === 'discount' ? '−' : '+';
                $unit = ['percent' => '%', 'per_kg' => '/кг', 'fixed' => ''][$r['calc_type']] ?? '';
                return $r['name'] . ' (' . $sign . Format::num($r['value']) . $unit . ')';
            },
        ];

        $e['faq'] = [
            'title' => '❓ FAQ',
            'one'   => 'Вопрос',
            'table' => 'faq',
            'order' => 'sort_order, id',
            'has_active' => true,
            'fields' => [
                ['name' => 'question', 'label' => 'Вопрос', 'type' => 'str'],
                ['name' => 'answer', 'label' => 'Ответ', 'type' => 'text'],
                ['name' => 'sort_order', 'label' => 'Порядок сортировки', 'type' => 'int'],
            ],
            'row' => fn(array $r) => (string) $r['question'],
        ];

        $e['whs'] = [
            'title' => '📍 Склады',
            'one'   => 'Склад',
            'table' => 'warehouses',
            'order' => 'sort_order, id',
            'has_active' => true,
            'fields' => [
                ['name' => 'name', 'label' => 'Название', 'type' => 'str'],
                ['name' => 'address', 'label' => 'Адрес', 'type' => 'text'],
                ['name' => 'note', 'label' => 'Примечание', 'type' => 'text', 'required' => false],
                ['name' => 'sort_order', 'label' => 'Порядок сортировки', 'type' => 'int'],
            ],
            'row' => fn(array $r) => (string) $r['name'],
        ];

        $e['cnt'] = [
            'title' => '☎ Контакты',
            'one'   => 'Контакт',
            'table' => 'contacts',
            'order' => 'sort_order, id',
            'has_active' => true,
            'fields' => [
                ['name' => 'title', 'label' => 'Название (Телефон, Telegram...)', 'type' => 'str'],
                ['name' => 'value', 'label' => 'Значение', 'type' => 'text'],
                ['name' => 'sort_order', 'label' => 'Порядок сортировки', 'type' => 'int'],
            ],
            'row' => fn(array $r) => $r['title'] . ': ' . $r['value'],
        ];

        $e['adm'] = [
            'title' => '👮 Администраторы',
            'one'   => 'Администратор',
            'table' => 'admins',
            'order' => 'id',
            'has_active' => true,
            'fields' => [
                ['name' => 'tg_id', 'label' => 'Telegram ID', 'type' => 'int'],
                ['name' => 'name', 'label' => 'Имя', 'type' => 'str', 'required' => false],
            ],
            'row' => fn(array $r) => ($r['name'] ?: 'админ') . ' (' . $r['tg_id'] . ')',
        ];

        $e['set'] = [
            'title' => '⚙️ Настройки',
            'one'   => 'Настройка',
            'table' => 'settings',
            'order' => 'id',
            'has_active' => false,
            'can_add' => false,
            'fields' => [
                ['name' => 'value', 'label' => 'Значение', 'type' => 'text'],
                ['name' => 'description', 'label' => 'Описание', 'type' => 'text', 'required' => false],
            ],
            'row' => fn(array $r) => $r['key'] . ' = ' . Format::trunc((string) $r['value'], 24),
        ];

        return $e;
    }

    /** @return array<string,mixed>|null */
    public static function get(string $key): ?array
    {
        return self::all()[$key] ?? null;
    }

    /** Значение поля описания с настройками по умолчанию. */
    public static function field(array $entity, int $index): ?array
    {
        $f = $entity['fields'][$index] ?? null;
        if ($f === null) {
            return null;
        }
        $f['required'] = $f['required'] ?? true;
        $f['hint'] = $f['hint'] ?? '';
        $f['choices'] = $f['choices'] ?? [];
        return $f;
    }
}
